substrate/client/consensus/beefy/src/worker.rs

```diff
@@ -458,13 +458,16 @@ where
 			notification.tree_route,
 		);
 
-		self.runtime
-			.runtime_api()
-			.beefy_genesis(notification.hash)
-			.ok()
-			.flatten()
-			.filter(|genesis| *genesis == self.persisted_state.pallet_genesis)
-			.ok_or(Error::ConsensusReset)?;
+		match self.runtime.runtime_api().beefy_genesis(notification.hash) {
+			Ok(Some(genesis)) if genesis != self.persisted_state.pallet_genesis =>
+				return Err(Error::ConsensusReset),
+			Ok(_) => {},
+			Err(api_error) => {
+				// This can happen in case the block was already pruned.
+				// Mostly after warp sync when finality notifications are piled up.
+				debug!(target: LOG_TARGET, "🥩 Unable to check beefy genesis: {}", api_error);
```

Comment:
Another option would be to store the `beefy_genesis` in `UnpinnedFinalityNotification`. We could retrieve it inside the transformer, when it should be available. Would that fix the issue ?

Reply:
> Another option would be to store the `beefy_genesis` in `UnpinnedFinalityNotification`. We could retrieve it inside the transformer, when it should be available. Would that fix the issue ?

This could work :+1: .

Regarding missing the ConsensusReset. So eventually we will catch up to the latest blocks and will find that the beefy_genesis is indeed different to what we have stored. In that case we would trigger the consensus reset, just later as we reach the tip. Is that acceptable? What are the risks here? If not acceptable we could add beefy genesis in the `UnpinnedFinalityNotification` as you proposed.

Reply:
Hmmm looked a bit on the code. I was thinking there might be problems with the authority set, or the payload, but it looks like this can't happen. I think it's ok. But I would also wait for @acatangiu 's input on this.

Reply:
We should still test it on Rococo where we have actually reset consensus and confirm we're not missing anything

Reply:
Ah okay I was not aware that this was done on rococo, will perform some tests there before merging